### ISSUE TITLE
Different result on groupby data

### ISSUE THREAD

**Issue by nirupamkar** [Microsoft/LightGBM#1705]
## Environment info

C++/Python/R version: Python 2.7

## Issue
I am using lightgbm2.1.0 for a classification problem. If I groupby the features and use the count as weight, ideally there should not be any change in the model as loss function should remain same. However I am seeing a change. Please let me know if I am missing something.

## Reproducible examples
```python
import pandas as panda
import numpy as numpy
from sklearn.metrics import roc_auc_score,log_loss

np.random.seed(42)
x=pd.DataFrame({"y":[int(np.random.random()>.8) for i in range(100)],
               "x1":[np.round(np.random.random(),1) for i in range(100)],"x2":[np.round(np.random.random(),1) for i in range(100)]})
x["weight"]=1
import lightgbm as lgb

valid_features=["x1","x2"]

lgb_train_data = lgb.Dataset(x[valid_features], label=x.y.values, 
                                  weight=x.weight.values,
                                  free_raw_data=False)
parameters = {
"boosting_type": 'gbdt', 
"objective" : 'binary',
    "n_estimators":10,
    "min_data":2,
    "random_state":2
}
print "running "
model2 = lgb.train(parameters,
lgb_train_data,
num_boost_round=2,
early_stopping_rounds=30,valid_sets=[lgb_train_data], 
                     valid_names=['train'])



log_loss(x.y,model2.predict(x[valid_features]))
0.34608254346789125


x1=x.groupby(["x1","x2","y"]).agg({"weight":"sum"}).reset_index()
lgb_train_data = lgb.Dataset(x1[valid_features], label=x1.y.values, 
                                  weight=x1.weight.values,
                                  free_raw_data=False)
parameters = {
"boosting_type": 'gbdt', 
"objective" : 'binary',
    "n_estimators":10,
    "min_data":2,
    "random_state":2
}
print "running "
model2 = lgb.train(parameters,
lgb_train_data,
num_boost_round=2,
early_stopping_rounds=30,valid_sets=[lgb_train_data], 
                     valid_names=['train'])

log_loss(x.y,model2.predict(x[valid_features]))
0.3561871134153139

```

**Comment by guolinke** [comment]
Aren't they the same in your pasted code ?

**Comment by nirupamkar** [comment]
yes, rectified it.

**Comment by guolinke** [comment]
I think the groupby will change the order of samples. 
Therefore, there may are some changes due to float point data sumup problems.

**Comment by nirupamkar** [comment]
But this leads to a massive performance drop in my actual scenario (this issue is an example). It is like 2 point drop in terms of AUC. Do you think float point data sumup problems can lead this significant change? Here also you can see the logloss in the groupby case got increased by around 0.01

**Comment by nirupamkar** [comment]
@guolinke please correct me if I am wrong.
I am seeing same observation in XGB also but it gets resolved if subsample=1 is used . So xgboost does not use stratified subsampling based on the row weight (confirmed ).
For lightgbm's case, even if subsample=1 is taken the issue is not resolved. Is that because in this case too lightgbm uses goss and it does not use weighted sampling ?

**Comment by guolinke** [comment]
@nirupamkar the goss isn't enabled by default. 
I think you can try set min_data=1, and using min_hessian instead.

**Comment by bbennett36** [comment]
Well the weights are different so certain instances in x1 will have more of an impact than x. You can see in the counts below -
```
x.weight.value_counts()
1    100
Name: weight, dtype: int64
```

```
x1.weight.value_counts()
1    54
2    15
3     4
4     1
Name: weight, dtype: int64
```

Also, if you print out x & x1 they are not in the same order which is also affecting it. However, if you add more iterations then the results are almost identical. (I always set num_iterations high with early stopping so I'm not stopping the model to early)

```
lgb_train_data = lgb.Dataset(x[valid_features], label=x.y.values, 
                                  weight=x.weight.values,
                                  free_raw_data=False)

parameters = {
"boosting_type": 'gbdt', 
"objective" : 'binary',
    "n_estimators":200,
    "min_data":2,
    "random_state":2,
    "boost_from_average": False
}
print("running")
model2 = lgb.train(parameters,
lgb_train_data,
num_boost_round=10,
early_stopping_rounds=30,valid_sets=[lgb_train_data], 
                     valid_names=['train'], verbose_eval=200)

log_loss(x.y,model2.predict(x[valid_features]))
```

```
[200]	train's binary_logloss: 0.174036
0.17403560480100944
```

```
x1=x.groupby(["x1","x2","y"]).agg({"weight":"sum"}).reset_index()
lgb_train_data = lgb.Dataset(x1[valid_features], label=x1.y.values, 
                                  weight=x1.weight.values,
                                  free_raw_data=False)
parameters = {
"boosting_type": 'gbdt', 
"objective" : 'binary',
    "n_estimators":200,
    "min_data":2,
    "random_state":2,
    "boost_from_average": False

}
print("running ")
model2 = lgb.train(parameters,
lgb_train_data,
num_boost_round=10,
early_stopping_rounds=30,valid_sets=[lgb_train_data], 
                     valid_names=['train'], verbose_eval=200)

log_loss(x.y,model2.predict(x[valid_features]))
```

```
[200]	train's binary_logloss: 0.174016
0.17401633362250837
```

**Comment by guolinke** [comment]
Thanks @bbennett36 
Another reason is related to `min_data`, which constraint the minimal data per leaf node.
Although the total weights of your data are not changed, the number of instances indeed changes. 
Therefore, you can try ```min_data=1`` to disable this constraint.
